src/qnetvo/network_nodes.py

```diff
@@ -0,0 +1,250 @@
+    :type num_out: int
+
+    :param wires: The wires on which the node operates.
+    :type wires: list[int]
+
+    :param cc_wires: The classical communication wires input to the node.
+    :type cc_wires: list[int]
+
+    :param cc_wires_out: The classical communication wires to output measurement results on.
+    :type cc_wires_out: list[int]
+
+    :param ansatz_fn: A `PennyLane quantum circuit function <https://docs.pennylane.ai/en/stable/introduction/circuits.html>`_
+        called either as ``circuit(settings, wires)``, or as ``circuit(settings, wires, cc_wires)`` where
+        ``settings`` is an *array[float]* of length ``num_settings`` and ``cc_wires`` contains the measurement results
+        received from upstream nodes.
+    :type ansatz_fn: function
+
+    :param num_settings: The number of settings parameterizing the ``ansatz_fn`` circuit.
+    :type num_settings: int
+
+    :returns: An instance of the ``NetworkNode`` class.
+
+    **Calling a Network Node's Ansatz Function:**
+
+    Given an instantiated network node, ``node = NetworkNode(*args)``, its ansatz quantum circuit functtion
+    can be called as ``node(settings, cc_wires)`` or, if ``node.cc_wires==[]``, the node can be called
+    as  ``node(settings)`` if ``cc_wires=[]``.
+
+    **Attributes:**
+
+    All inputs are stored as class attributes under the same name, *e.g.*, the ``num_in`` argument
+    is stored as ``node.num_in``.
+    """
+
+    def __init__(
+        self,
+        num_in=1,
+        num_out=1,
+        wires=[],
+        cc_wires=[],
+        cc_wires_out=[],
+        ansatz_fn=None,
+        num_settings=0,
+    ):
+        self.num_in = num_in
+        self.num_out = num_out
+
+        self.wires = wires
+        self.cc_wires = cc_wires
+        self.cc_wires_out = cc_wires_out
+
+        self.ansatz_fn = ansatz_fn if ansatz_fn else self.default_ansatz_fn
+        self.num_settings = num_settings
+
+    def __call__(self, settings=[], cc_wires=[]):
+        args = [settings, self.wires]
+        if cc_wires:
+            args += [cc_wires]
+
+        return self.ansatz_fn(*args)
+
+    def default_ansatz_fn(self, settings, wires, cc_wires=[]):
+        pass
+
+
+class NoiseNode(NetworkNode):
+    """A network node that applies noise to its local qubit wires.
+
+    We model noise to be independent from classical inputs and upstream measurement results.
+    Therefore, only the ``wires`` and ``ansatz_fn`` attributes can be set.
+    All inputs and attributes are inherited from the :class:`qnetvo.NetworkNode` class.
+
+    The ``ansatz_fn`` for a ``NoiseNode`` should take the following form:
+
+    .. code-block:: python
+
+        def noise_ansatz(settings, wires):
+            # apply noise operation
+            qml.Depolarizing(0.5, wires=wires[0])
+            qml.AmplitudeDamping(0.5, wires=wires[1])
+
+    Note that the ``settings`` parameter is not typically used in a noise node.
+
+    :returns: An instantiated ``NoiseNode`` class.
```

Comment:
That is correct. Although, I can see the case for copy-pasting the relevant fields to each node. As it currently stands, I can see how it may be confusing.

Reply:
I modified such that the parameters for each node are defined explicitly.